web/server/codechecker_server/permissions.py

```diff
@@ -293,21 +293,51 @@ def __get_perm_record(self, auth_name, is_group):
                 query(SysPerm). \
                 filter(and_(
                     SysPerm.permission == self._perm_name,
-                    SysPerm.name == auth_name,
+                    func.lower(SysPerm.name) == auth_name.lower(),
                     SysPerm.is_group == is_group
                 )).one_or_none()
 
             return record
 
+        def __get_stored_auth_name_and_permissions(self, auth_name, is_group):
+            SysPerm = config_db_model.SystemPermission
+
+            stored_auth_name = auth_name
+            permission_set = set()
+            for name, permission in self.__session.query(
+                    SysPerm.name, SysPerm.permission).filter(and_(
+                    func.lower(SysPerm.name) == auth_name.lower(),
+                    SysPerm.is_group == is_group)):
+                stored_auth_name = name
+                permission_set.add(permission)
+
+            return (stored_auth_name, permission_set)
+
         def _add_perm_impl(self, auth_name, is_group=False):
-            perm_record = self.__get_perm_record(auth_name, is_group)
-            if perm_record is None:
-                perm_record = config_db_model.SystemPermission(
+            if not auth_name:
```

Comment:
This function is very similar to the function which can be found in the `ProductPermission` class. The only difference is how we create new permission. I think we can use inheritance to avoid code duplication similar what I mentioned above.

Please also add docstrings for functions.

Reply:
Yes, you are right. But it seems all member functions duplicated. So I did not do a whole refactor, I added two functions only following the existing code style.
I wrote docstrings for new functions.